pkg/writer/junit_xml.go

```diff
@@ -0,0 +1,192 @@
+type JUnitSkipMessage struct {
+	Message string `xml:"message,attr"`
+}
+
+// JUnitProperty represents a key/value pair used to define properties.
+type JUnitProperty struct {
+	Name  string `xml:"name,attr"`
+	Value string `xml:"value,attr"`
+}
+
+// JUnitFailure contains data related to a failed test.
+type JUnitFailure struct {
+	Message  string `xml:"message,attr"`
+	Type     string `xml:"type,attr"`
+	Contents string `xml:",chardata"`
+}
+
+func newJunitTestSuites(summary results.ScanSummary) JUnitTestSuites {
+	return JUnitTestSuites{
+		Tests:    summary.TotalPolicies,
+		Name:     testSuitesName,
+		Failures: summary.ViolatedPolicies,
+		Time:     fmt.Sprint(summary.TotalTime),
+	}
+}
+
+func newJunitTestSuite(summary results.ScanSummary) JUnitTestSuite {
+	return JUnitTestSuite{
+		Name:     testSuiteName,
+		Tests:    summary.TotalPolicies,
+		Time:     fmt.Sprint(summary.TotalTime),
+		Failures: summary.ViolatedPolicies,
+		Package:  summary.ResourcePath,
+		Properties: []JUnitProperty{
+			{
+				Name:  "Terrascan Version",
+				Value: version.Get(),
+			},
+		}}
+}
+
+func init() {
+	RegisterWriter(junitXMLFormat, JUnitXMLWriter)
+}
+
+// JUnitXMLWriter writes scan summary in junit xml format
+func JUnitXMLWriter(data interface{}, writer io.Writer) error {
+	output, ok := data.(policy.EngineOutput)
+	if !ok {
+		return fmt.Errorf("incorrect input for JunitXML writer, supportted type is policy.EngineOutput")
+	}
+
+	junitXMLOutput := convert(output)
+
+	return XMLWriter(junitXMLOutput, writer)
+}
+
+// convert is helper func to convert engine output to JUnitTestSuites
+func convert(output policy.EngineOutput) JUnitTestSuites {
+	testSuites := newJunitTestSuites(output.Summary)
+	// since we have a single suite for now, a suite will have same data as in root level element testsuites
+	suite := newJunitTestSuite(output.Summary)
+
+	tests := violationsToTestCases(output.ViolationStore.Violations, false)
+	if tests != nil {
+		suite.TestCases = append(suite.TestCases, tests...)
+	}
+
+	skippedTests := violationsToTestCases(output.ViolationStore.SkippedViolations, true)
+	if skippedTests != nil {
+		suite.TestCases = append(suite.TestCases, skippedTests...)
+	}
+
+	testSuites.Suites = append(testSuites.Suites, suite)
+
+	return testSuites
+}
+
+// violationsToTestCases is helper func to convert scan violations to JunitTestCases
+func violationsToTestCases(violations []*results.Violation, isSkipped bool) []JUnitTestCase {
+	testCases := make([]JUnitTestCase, 0)
+	for _, v := range violations {
+		var testCase JUnitTestCase
+		if isSkipped {
+			testCase = JUnitTestCase{Failure: new(JUnitFailure), SkipMessage: new(JUnitSkipMessage)}
+		} else {
+			testCase = JUnitTestCase{Failure: new(JUnitFailure)}
+		}
+		testCase.Classname = v.File
+		testCase.Name = "[ERROR] resource: " + fmt.Sprintf(`"%s"`, v.ResourceName) + " at line: " + fmt.Sprintf("%d", v.LineNumber) + " violates: " + v.RuleID
```

Comment:
We can wrap this whole string under fmt.Sprintf

Reply:
skipMsg := new(JUnitSkipMessage)
skipMsg.Message = v.Comment

testCase = JUnitTestCase{Failure: new(JUnitFailure), SkipMessage: skipMsg}